Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -464,11 +470,80 @@ public CsharpCompiler(ErrorReporter reporter)
 
       CompileDatatypeDestructorsAndAddToInterface(dt, wr, interfaceTree, DtT_TypeArgs);
 
+      DtDowncastClone(dt, interfaceTree, nonGhostTypeArgs, toInterface: true);
+      if (!dt.IsRecordType) {
+        DtDowncastClone(dt, wr, nonGhostTypeArgs);
+      }
+
       CompileDatatypeInterfaceMembers(dt, interfaceTree);
 
       return new ClassWriter(this, btw, null);
     }
 
+    private void DtDowncastClone(DatatypeDecl dt, ConcreteSyntaxTree wr, List<TypeParameter> nonGhostTypeArgs, bool toInterface = false, bool lazy = false, DatatypeCtor ctor = null) {
+      if (nonGhostTypeArgs.Any(ty => ty.Variance == TypeParameter.TPVariance.Contra)
+          || dt.Ctors.Any(ctor => ctor.Formals.Any(f => f.Type.IsArrowType || f.Type.IsRefType))) {
+        return;
+      }
+      var typeArgs = TypeParameters(nonGhostTypeArgs, safe: true);
+      var dtArgs = "_I" + dt.CompileName + typeArgs;
+      string PrintConverter((TypeParameter, int) t) {
+        var (tArg, i) = t;
+        var name = IdName(tArg);
+        return $"Func<{name}, __{name}> converter{i}";
+      }
+      var argsWithIndex = nonGhostTypeArgs.Zip(Enumerable.Range(0, nonGhostTypeArgs.Count));
+
+      if (!toInterface) {
+        wr.Write($"public {(ctor != null || lazy ? (dt.IsRecordType ? "" : "override ") : "abstract ")}");
+      }
+
+      wr.Write($"{dtArgs} DowncastClone{typeArgs}({argsWithIndex.Comma(PrintConverter)})");
```

Comment:
Comma has an overload that also sends the `index` to the function, so you can use `nonGhostTypeArgs` instead of `argsWithIndex`

Reply:
Why was this resolved?

Reply:
You asked me to use the overloaded version and I did.

Reply:
Okay, maybe I looked at an outdated version then. Sorry if I did.